### PR TITLE
Intermediate format to simplify syncing

### PR DESCRIPTION
**Pull request by kirkbyo**
Closes https://github.com/andymatuschak/orbit/issues/260

## What
This PR defines a new package called `Ingester` which outlines an intermediate file format to sync prompts to an `OrbitStore`.

## Why
Having this intermediate file format allows for the decoupling between source parsing and syncing. 

## What is the format?
```typescript
export type Ingestible = {
  sources: IngestibleSource[];
};

export interface IngestibleSource {
  // unique identifier representing a source which tasks might share, used for clustering; can be a URL or an arbitrary UUID
  identifier: IngestibleSourceIdentifier;
  // title that can be used when displaying the source.
  title: string;
  // prompts associated with the given source
  prompts: IngestiblePrompt[];
  // an optional URL to open when the user indicates they'd like to navigate to the provenance of the task
  url?: string;
  // an optional const that can be used to customize the visual appearance of the source during review
  colorPaletteName?: ColorPaletteName;
}
```

In this schema, producers (Anki, Bear, General Markdown, Biome) generate an "ingestible" file which are associated with an arbitrary number of sources. Prompts are associated with sources. Bear might produce a source per note, while Anki might produce a source per deck. A producer generates a stable identifier for each source. The ingester package is responsible for diffing between the prompts in the Orbit store, and the ones provided by the producer.

I believe this current format strikes a good balance between the "inferred" vs. "explicit" problem that I attempted to enumerate in the original issue. In this model, a producer does not need to output the absolute set of prompts. It can instead only produce sources that need to be synced. Sources which are not provided, are **not considered**. See `ingest.test.ts` test cases which I think capture the expected usage of this API.

## How has this been tested?
I am currently working on porting the `note-sync` package to use this code on https://github.com/kirkbyo/orbit/commits/bear-import. That work should help give some confidence in that this API is sound. That being said, I do not have a large corpus of markdown files with embedded prompts (and not a Bear user at that), so if you have any ideas on how I could test this further let me know.

### CODE REVIEW

**Comment by kirkbyo** [comment]
@andymatuschak This PR is in a good state in case you wanted to give it a general first pass review for any major pitfalls that I might have overlooked!

Remaining todo's on my end:

- [ ] Support cloze prompts
- [ ] Support individual prompts having an `openURL`
- [ ] Finish migrating the `note-sync` package to use this package in a separate branch. This will help ensure that the current implementation is sound e2e _(note: the actual migration work will be reviewed in a separate PR)_.

File: packages/store-fs/src/sqlite.ts [hunk 1]
```diff
@@ -44,10 +44,13 @@ export class SQLDatabaseBackend implements DatabaseBackend {
 
   static attachmentURLProtocol = "x-orbit-sqlattachment";
 
-  constructor(path: string) {
+  constructor(path: string, options?: { enableDebugLogs?: boolean }) {
```
Comment (kirkbyo) [inline packages/store-fs/src/sqlite.ts:47]: these logs were becoming quite noisy in the unit tests so I created a flag to disable them. Happy to revert these changes and just stub out `console.log` in my tests if you would prefer.

The more SOLID approach would to pass in a `Logger` interface, but that felt a bit heavy for this case.

File: packages/ingester/package.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@withorbit/ingester",
```
Comment (kirkbyo) [inline packages/ingester/package.json:2]: Naming is still up in the air, let me know if you feel if something else might be more apt!

I also didn't license this package, let me know if you would like me to port over any of the existing licenses into this package.

File: packages/ingester/src/ingestible.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  __sourceIDOpaqueType: never;
+};
+
+export interface IngestibleSource {
+  // unique identifier representing a source which tasks might share, used for clustering; can be a URL or an arbitrary UUID
+  identifier: IngestibleSourceIdentifier;
+  // title that can be used when displaying the source.
+  title: string;
+  // prompts associated with the given source
+  prompts: IngestiblePrompt[];
```
Comment (andymatuschak) [inline packages/ingester/src/ingestible.ts:21]: Two somewhat invasive suggestions; feel free to take or leave.

**Prompt -> TaskSpec**
I think it might be more parsimonious to replace `prompts` with `taskSpecs` here? I think `TaskSpec` is the general form of what you're pointing to here. It would let you avoid redefining QA prompts below and mapping between tasks and prompts. And once we define non-memory tasks (which I do intend to do!) I expect to want to ingest them from Markdown in a similar way.

A very simple task type I expect to add would simply display some content, and then provide some controls which let you adjust the subsequent interval. I might use this to show myself visual inspiration, or for tasks like "sing a minor sixth upward", etc. I can imagine wanting to ingest these from my Markdown notes by tagging any block with `#orbitTask`.

Feel free to YAGNI me.

**Provenance**
For that matter, the other properties seem mostly duplicative of `TaskProvenance`. Could this whole thing become?:

```
export interface IngestibleSource {
   // unique identifier representing a source which tasks might share, used for clustering; can be a URL or an arbitrary UUID
   identifier: IngestibleSourceIdentifier;

   specs: TaskSpec[];
   provenance: TaskProvenance;
}

File: packages/ingester/src/ingest.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+  for (const source of sources) {
+    if (groupedEntities[source.identifier]) {
+      // determine which prompts are new and which have been deleted
+      const existingPrompts = [...groupedEntities[source.identifier]];
+
+      for (const prompt of source.prompts) {
+        // NOTE: this `findIndex` is a bit expensive since it scans of each iteration.
+        // If this becomes a bottleneck we should precompute a hash for each task prompt.
+        const existingPromptIndex = existingPrompts.findIndex(
```
Comment (andymatuschak) [inline packages/ingester/src/ingest.ts:43]: This suggestion complicates your design here, but I figure it's worth discussing:

In the present design, ingestion sources surface a bunch of prompt content (i.e. front/back strings). The shared ingestion machinery implements a notion of "identity" on top of this by matching that verbatim content with existing prompt content.

This is expensive, but it might also be more limiting than it needs to be. I implemented that policy for the Markdown importer because that format has no way of tracking identity over time. But Biome could very reasonably make a little ID for each prompt you create in its UI! We could be using those IDs to maintain a prompt's identity even when its text is modified.

Concretely, an alternative would be to have `IngestionPrompt` include an arbitrary identifier, provided by the source. For Markdown prompts, that'd just be a hash of the content; for other sources it might be some stable identifier. We'd store that in a metadata key on the `Task`. Then we could replace this double-loop with set arithmetic.

(Feel free to YAGNI me of course)

**Review by andymatuschak — COMMENTED** [review]
Exciting!! I made a few high-level suggestions for discussion; I didn't finish looking through in detail because things will probably change if you choose to follow any of them. But feel free to reject all these and continue directly—I don't want to complicate. Just let me know what you'd prefer!

**Review by kirkbyo — COMMENTED** [review]
Thanks for the first pass! This PR should be good to now.

A couple things to tackle in follow up PRs:
- I have opened https://github.com/andymatuschak/orbit/pull/268 which migrates the note-sync to use this work, and is also how I tested this code e2e to ensure it is working. That being said, I do not have a corpus of markdown notes with prompts so I can't assess the performance of this work.
- Now that task items have an identifier associated that is stored in prompt metadata, we will need a migration script to allow existing users of the `note-sync` package to migrate to the interpreter above.
- Provenance updates and moves are currently not tracked in this work. Implementing this behaviour is relatively straightforward with the current setup, but I didn't want to complicate this review any further.

File: packages/ingester/src/ingest.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  }
+  return mapping;
+}
+
+function groupEntitiesBySourceIdentifiers(entities: Task<TaskContent>[]) {
+  // [01/05/22]: our current versin of typescript does not allow
+  // `IngestibleItemIdentifier` to be the key of an object. Newer
+  // releases of typescript support this behavior.
+  // https://github.com/microsoft/TypeScript/pull/44512
+  const mapping: Record<string, Task<TaskContent>> = {};
```
[the remark this answers is not in the document]
Comment (andymatuschak) [reply packages/ingester/src/ingest.ts:104]: Unfortunate, but sounds like the right call.

File: packages/ingester/package.json (hunk 2 above)
Comment (andymatuschak) [reply packages/ingester/package.json:2]: `ingester` is fine for now. It'd be great if you could copy the Apache `LICENSE` from one of the other library packages into the root here.

File: packages/ingester/src/ingest.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+) {
+  const events: Event[] = [];
+
+  // TODO: create a new query or extend the entity query such that
+  // we can filter on only sources that would have used this ingester
+  const query: DatabaseEntityQuery<Task> = {
+    entityType: EntityType.Task,
+  };
+  const entities = await store.database.listEntities(query);
+  const groupedEntities = groupEntitiesByProvenanceIdentifiers(entities);
```
Comment (andymatuschak) [inline packages/ingester/src/ingest.ts:43]: Reading this, I noticed that I had a lot of trouble following these different maps: what's being mapped to what? One thing I liked about Cocoa was its very consistent naming conventions—meant less thinking while coding, since names were largely formulaic, but also helped legibility. Maps/dictionaries/records used the convention `valuesByKey`, so the three maps here might be named: `existingEntitiesByProvenanceIdentifiers` (I added "existing" here because that distinction is important in this routine), `existingEntitiesByItemIdentifiers`, and `ingestibleItemsByItemIdentifiers`.

No obligation to change this, just thought you might find it interesting.

File: packages/ingester/src/ingest.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  const timeMillis = opts.ingestDateMilis;
+
+  for (const source of sources) {
+    if (groupedEntities[source.identifier]) {
+      // determine which prompts are new and which have been deleted
+      const existingPrompts = mapEntitiesByItemIdentifiers(
```
Comment (andymatuschak) [inline packages/ingester/src/ingest.ts:49]: Prompts? Tasks? Entities?

File: packages/ingester/src/ingest.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    if (mapping[provenance.identifier]) {
+      mapping[provenance.identifier].push(entity);
+    } else {
+      mapping[provenance.identifier] = [entity];
+    }
+  }
+  return mapping;
+}
+
+function mapEntitiesByItemIdentifiers(entities: Task<TaskContent>[]) {
```
Comment (andymatuschak) [inline packages/ingester/src/ingest.ts:99]: Thanks for renaming. :)

File: packages/ingester/src/validateIngestible.test.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import { createIngestibleValidator } from "./validateIngestible";
+
+it("accepts valid schema", () => {
+  const validator = createIngestibleValidator({
+    mutateWithDefaultValues: true,
+  });
+  const { isValid, errors } = validator.validate({
+    sources: [
+      {
+        identifier: "aaaaaaaaaaaaaaaaaaaaaa",
```
[the remark this answers is not in the document]
Comment (andymatuschak) [reply packages/ingester/src/validateIngestible.test.ts:10]: Recapping our phone discussion: no need to validate these external IDs; we don't/can't know what scheme they're using.

**Review by andymatuschak — APPROVED** [review]
Looking great! Some comments throughout, but feel free to take/leave all. This is ready when you're ready.

File: packages/ingester/src/ingest.ts (hunk 6 above)
Comment (andymatuschak) [reply packages/ingester/src/ingest.ts:43]: Ah, the naming is complicated slightly by the fact that the first map is actually of _groups_; if this were Cocoa I'd probably name it `existingEntityGroupsByProvenanceIdentifiers` (of course, Cocoa likes long variable names, and that's perhaps less idiomatic in JS culture)

File: packages/ingester/src/ingest.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        createIngestTaskForSource(source, timeMillis),
+      );
+      events.push(...ingestEvents);
+    }
+  }
+  return events;
+}
+
+function groupEntitiesByProvenanceIdentifiers(entities: Task<TaskContent>[]) {
+  const mapping: Record<TaskProvenance["identifier"], Task<TaskContent>[]> = {};
```
[the remark this answers is not in the document]
Comment (kirkbyo) [reply packages/ingester/src/ingest.ts:85]: Oh TIL! I didn't realize JS had a map construct. That is great!

**Comment by kirkbyo** [comment]
Thanks for the review @andymatuschak! I believe I addressed all the comments and also added two new error conditions: duplicate item identifiers and missing item identifier in the metadata. I am going to go ahead and merge this so that the other PR's can be rebased 🚀
